images/converter/default.go

```diff
@@ -60,41 +60,97 @@ type ConvertHooks struct {
 
 // IndexConvertFuncWithHook is the convert func used by Convert with hook functions support.
 func IndexConvertFuncWithHook(layerConvertFunc ConvertFunc, docker2oci bool, platformMC platforms.MatchComparer, hooks ConvertHooks) ConvertFunc {
-	c := &defaultConverter{
+	return newDefaultWithOpts(convertOpts{
 		layerConvertFunc: layerConvertFunc,
 		docker2oci:       docker2oci,
 		platformMC:       platformMC,
-		diffIDMap:        make(map[digest.Digest]digest.Digest),
 		hooks:            hooks,
+	}).Convert
+}
+
+func NewDefault(opts ...Opt) (*DefaultConverter, error) {
+	var copts convertOpts
+	for _, o := range opts {
+		if err := o(&copts); err != nil {
+			return nil, err
+		}
+	}
+
+	return newDefaultWithOpts(copts), nil
+}
+
+func newDefaultWithOpts(copts convertOpts) *DefaultConverter {
+	if copts.platformMC == nil {
+		copts.platformMC = platforms.All
+	}
+
+	return &DefaultConverter{
+		convertOpts: copts,
+	}
+}
+
+type DefaultConverter struct {
+	convertOpts
+}
+
+type diffIDMap struct {
+	diffIDs map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
+	mutex   sync.RWMutex
+}
+
+func makeRefKey(desc ocispec.Descriptor) (string, error) {
+	key, err := json.Marshal(desc)
+	if err != nil {
+		return "", err
+	}
+	return string(key), nil
+}
+
+// singleflightDo makes sure that only one conversion execution is
+// in-flight for a given descriptor at a time.
+func (c *DefaultConverter) singleflightDo(ctx context.Context, cs content.Store, desc ocispec.Descriptor, dm *diffIDMap, convert convertFuncWithDiffMap) (*ocispec.Descriptor, error) {
+	if c.sg == nil {
+		return convert(ctx, cs, desc, dm)
+	}
+
+	key, err := makeRefKey(desc)
+	if err != nil {
+		return nil, err
 	}
-	return c.convert
+
+	newDesc, err, _ := c.sg.Do(key, func() (interface{}, error) {
+		return convert(ctx, cs, desc, dm)
```

Comment:
It seems that `*diffIDMap` won't be updated if this call doesn't actually run `convert` because of another one running? This will result in the config blob being broken.

Reply:
@ktock Thanks for pointing out, I realized that currently config is usually only updated by `convertManifest`, so `diffIDMap` only need to be passed to `convertConfig`.